### PR TITLE
Make Windows build C++17 compliant

### PR DESCRIPTION
**Pull request by anmyachev**

### CODE REVIEW

File: third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp [hunk 1]
```diff
@@ -117,15 +117,8 @@ getValuesFromBlockPointerStruct(Value blockPointerStruct,
       blockPointerStruct.getLoc(), blockPointerStruct, rewriter);
   assert(elems.size() == 7 &&
          "unexpected number of values unpacked from a block pointer");
-  BlockPointerValues values{
```
Comment (anmyachev) [inline third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp:120]: This code was according to `c++20` standard.

File: CMakeLists.txt [hunk 2]
```diff
@@ -58,15 +60,13 @@ endif()
...
-  set(CMAKE_CXX_STANDARD 17)
   set(CMAKE_C_FLAGS_TRITONRELBUILDWITHASSERTS "-O2 -g")
   set(CMAKE_CXX_FLAGS_TRITONRELBUILDWITHASSERTS "-O2 -g")
   set(CMAKE_C_FLAGS_TRITONBUILDWITHO1 "-O1")
   set(CMAKE_CXX_FLAGS_TRITONBUILDWITHO1 "-O1")
 else()
-  set(CMAKE_CXX_STANDARD 20)
-  set(CMAKE_C_FLAGS_TRITONRELBUILDWITHASSERTS "/Zi /RTC1 /bigobj /Zc:preprocessor")
-  set(CMAKE_CXX_FLAGS_TRITONRELBUILDWITHASSERTS "/Zi /RTC1 /bigobj /Zc:preprocessor")
+  set(CMAKE_C_FLAGS_TRITONRELBUILDWITHASSERTS "/Zi /RTC1 /bigobj /Zc:preprocessor /permissive-")
```
Comment (anmyachev) [inline CMakeLists.txt:68]: Ref: https://learn.microsoft.com/en-us/cpp/build/reference/permissive-standards-conformance?view=msvc-170

File: third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp [hunk 3]
```diff
@@ -117,15 +117,8 @@ getValuesFromBlockPointerStruct(Value blockPointerStruct,
...
-      .base = elems[6],
-      .baseWidth = elems[3],
-      .baseHeight = elems[2],
-      .rowStride = elems[4],
-      .colStride = elems[5],
-      .offsetBaseX = elems[1],
-      .offsetBaseY = elems[0],
-  };
+  BlockPointerValues values{elems[6], elems[3], elems[2], elems[4],
```
Comment (pbchekin) [inline third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp:120]: This is not readable and maintainable as the previous code.

Comment (anmyachev) [reply third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp]: I added comments, this is the best that can be done in c++17 standard, if I'm not mistaken.

**Review by whitneywhtsang — COMMENTED** [review]
Do you know if upstream is opened to changing to C++20 compliant?

**Comment by anmyachev** [comment]
> Do you know if upstream is opened to changing to C++20 compliant?

I don't know, it seems to me that there is a connection with PyTorch, which still uses `c++17` standard. I suppose that as soon as it updates the standard, Triton will do the same.

**Review by whitneywhtsang — COMMENTED** [review]
Adding Victor as a reviewer to see if he has any other ideas, otherwise this change LGTM.

File: third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp [hunk 4]
```diff
@@ -117,15 +117,13 @@ getValuesFromBlockPointerStruct(Value blockPointerStruct,
...
-  BlockPointerValues values{
-      .base = elems[6],
-      .baseWidth = elems[3],
-      .baseHeight = elems[2],
-      .rowStride = elems[4],
-      .colStride = elems[5],
-      .offsetBaseX = elems[1],
-      .offsetBaseY = elems[0],
-  };
+  BlockPointerValues values{/*base*/ elems[6],
```
Comment (victor-eds) [inline third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp:120]:
```suggestion
  BlockPointerValues values{/*base=*/elems[6],
```

And same for other rows. `clang-tidy` likes that better.

**Comment by victor-eds** [comment]
LGTM but NIT.

File: third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp [hunk 5]
```diff
@@ -117,15 +117,13 @@ getValuesFromBlockPointerStruct(Value blockPointerStruct,
...
+                            /*baseWidth*/ elems[3],
+                            /*baseHeight*/ elems[2],
+                            /*rowStride*/ elems[4],
+                            /*colStride*/ elems[5],
+                            /*offsetBaseX*/ elems[1],
+                            /*offsetBaseY*/ elems[0]};
```
Comment (anmyachev) [inline third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp:126]:
```suggestion
                            /*baseWidth=*/ elems[3],
                            /*baseHeight=*/ elems[2],
                            /*rowStride=*/ elems[4],
                            /*colStride=*/ elems[5],
                            /*offsetBaseX=*/ elems[1],
                            /*offsetBaseY=*/ elems[0]};
```

**Review by victor-eds — APPROVED** [review]

**Pull request merged**